src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -0,0 +1,76 @@
+from collections import defaultdict
+
+import torch as th
+
+from poprox_concepts import Article, ArticleSet, InterestProfile
+from poprox_recommender.diversifiers import Calibrator, compute_kl_divergence
+from poprox_recommender.lkpipeline import Component
+from poprox_recommender.topics import extract_locality, normalized_topic_count
+
+
+# Locality Calibration uses MMR
+# to rerank recommendations according to
+# locality calibration
+class LocalityCalibrator(Component, Calibrator):
+    def __init__(self, theta: float = 0.1, num_slots=10):
+        super.__init__(theta, num_slots)
+
+    def __call__(self, candidate_articles: ArticleSet, interest_profile: InterestProfile) -> ArticleSet:
+        normalized_locality_prefs = normalized_topic_count(interest_profile.click_locality_counts)
+
+        if candidate_articles.scores is not None:
+            article_scores = th.sigmoid(th.tensor(candidate_articles.scores))
+        else:
+            article_scores = th.zeros(len(candidate_articles.articles))
+
+        article_scores = article_scores.cpu().detach().numpy()
+
+        article_indices = self.calibration(
+            article_scores,
+            candidate_articles.articles,
+            normalized_locality_prefs,
+            self.theta,
+            topk=self.num_slots,
+        )
+        return ArticleSet(articles=[candidate_articles.articles[int(idx)] for idx in article_indices])
+
+    def calibration(self, relevance_scores, articles, preferences, theta, topk) -> list[Article]:
```

Comment:
This method looks pretty much the same as the corresponding method on the `TopicCalibrator`. Wonder if it could move into the base class?

Reply:
Yeah that's certainly possible, I was bit hesitant before since the `calibration` function will call overridden `add_article_to_categories()` function, but maybe that's OK as well. I'll try factor `calibration()` and `normalized_categories_with_candidate()` into the parent class and see how it goes!